packages/react-composites/src/composites/CallWithChatComposite/adapter/AzureCommunicationCallWithChatAdapter.ts

```diff
@@ -760,38 +819,138 @@ export class AzureCommunicationCallWithChatAdapter implements CallWithChatAdapte
-        this.chatAdapter.off('participantsAdded', listener);
+        this.chatAdapterPromise.then((adapter) => {
+          adapter.off('participantsAdded', listener);
+        });
         break;
       case 'chatParticipantsRemoved':
-        this.chatAdapter.off('participantsRemoved', listener);
+        this.chatAdapterPromise.then((adapter) => {
+          adapter.off('participantsRemoved', listener);
+        });
         break;
       case 'callError':
         this.callAdapter.off('error', listener);
         break;
       case 'chatError':
-        this.chatAdapter.off('error', listener);
+        this.chatAdapterPromise.then((adapter) => {
+          adapter.off('error', listener);
+        });
         break;
       default:
         throw `Unknown AzureCommunicationCallWithChatAdapter Event: ${event}`;
     }
   }
 }
 
+/**
+ * Provides a way to get the chat thread ID for a given locator.
+ *
+ * @public
+ */
+export interface ChatThreadProvider {
+  getChatThread(): Promise<string>;
+}
+
+/**
+ * Arguments for use in {@link createAzureCommunicationCallWithChatAdapter} to join a Group Call with an associated Chat thread.
+ * @public
+ */
+export class CallAndChatProvider implements ChatThreadProvider {
+  public locator: CallAndChatLocator;
+
+  constructor(locator: CallAndChatLocator) {
+    this.locator = locator;
+  }
+
+  public async getChatThread(): Promise<string> {
+    return this.locator.chatThreadId;
+  }
+}
+
+/**
+ * Arguments for use in {@link createAzureCommunicationCallWithChatAdapter} to join a Teams meeting with an associated Chat thread.
+ *
+ * @public
+ */
+export class TeamsMeetingLinkProvider implements ChatThreadProvider {
+  public locator: TeamsMeetingLinkLocator;
+
+  constructor(locator: TeamsMeetingLinkLocator) {
+    this.locator = locator;
+  }
+
+  public async getChatThread(): Promise<string> {
+    return getChatThreadFromTeamsLink(this.locator.meetingLink);
+  }
+}
+
+/** @conditional-compile-remove(meeting-id) */
+/**
+ * Arguments for use in {@link createAzureCommunicationCallWithChatAdapter} to join a Teams meeting using meeting id.
+ *
+ * @public
+ */
+export class TeamsMeetingIdProvider implements ChatThreadProvider {
+  public locator: TeamsMeetingIdLocator;
+  private callAdapter: Promise<CallAdapter>;
+
+  constructor(locator: TeamsMeetingIdLocator, callAdapter: Promise<CallAdapter>) {
+    this.locator = locator;
+    this.callAdapter = callAdapter;
+  }
+
+  /**
+   * Wait call to be connected to get thread ID.
+   * @returns the chat thread ID for the given meeting ID.
+   */
+  public async getChatThread(): Promise<string> {
+    return new Promise<string>((resolve) => {
+      const stateChangeListener = (state: CallAdapterState): void => {
+        if (state.call?.state === 'Connected' && state.call.info?.threadId) {
+          resolve(state.call.info?.threadId);
+        }
+      };
+      this.callAdapter.then((adapter) => {
+        adapter.onStateChange(stateChangeListener);
```

Comment:
This needs refactoring, this is subscribing a new listener every time the `getChatThread` function is called, there is no equivalent `adapter.offStateChange` here